lnrpc/routerrpc/router_server.go

```diff
@@ -401,11 +456,299 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+			refHopHint.FeeProportionalMillionths =
+				rhLastHop.FeeProportionalMillionths
+
+			refHopHint.CLTVExpiryDelta = rhLastHop.CLTVExpiryDelta
+		}
+
+		// As long as we could be dealing with an LSP, strip off the
+		// last hop hint, since this is the LSP node which we will probe
+		// separately.
+		if len(rh) > 1 {
+			adjustedRouteHints = append(
+				adjustedRouteHints, rh[:len(rh)-1],
+			)
+		}
+	}
+
+	return true, adjustedRouteHints, &refHopHint
+}
+
+// probePaymentStream is a custom implementation of the grpc.ServerStream
+// interface. It is used to send payment status updates to the caller on the
+// stream channel.
+type probePaymentStream struct {
+	Router_SendPaymentV2Server
+
+	stream chan *lnrpc.Payment
+	ctx    context.Context //nolint:containedctx
+}
+
+// Send sends a payment status update to a payment stream that the caller can
+// evaluate.
+func (p *probePaymentStream) Send(response *lnrpc.Payment) error {
+	p.stream <- response
+
+	return nil
+}
+
+// Context returns the context of the stream.
+func (p *probePaymentStream) Context() context.Context {
+	return p.ctx
+}
+
+// sendProbePayment sends a payment to a target node in order to obtain
+// potential routing fees for it. The payment request has to contain a payment
+// hash that is guaranteed to be unknown to the target node, so it cannot settle
+// the payment. This method invokes a payment request loop in a goroutine and
+// awaits payment status updates.
+func (s *Server) sendProbePayment(ctx context.Context,
+	req *SendPaymentRequest) (*RouteFeeResponse, error) {
+
+	// We'll launch a goroutine to send the payment probes.
+	errChan := make(chan error, 1)
+	defer close(errChan)
+
+	paymentStream := &probePaymentStream{
+		stream: make(chan *lnrpc.Payment),
+		ctx:    ctx,
+	}
+	go func() {
+		err := s.SendPaymentV2(req, paymentStream)
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	// paymentTimeLockAndFee returns the fee and total time lock of the last
+	// attempt of the payment.
+	paymentTimeLockAndFee := func(p *lnrpc.Payment) (int64, int64) {
+		if len(p.Htlcs) > 0 {
+			lastAttempt := p.Htlcs[len(p.Htlcs)-1]
+			if lastAttempt.Route != nil {
+				return lastAttempt.Route.TotalFeesMsat,
+					int64(lastAttempt.Route.TotalTimeLock)
+			}
+		}
+
+		return 0, 0
+	}
+
+	for {
+		select {
+		case payment := <-paymentStream.stream:
+			switch payment.Status {
+			case lnrpc.Payment_INITIATED:
+			case lnrpc.Payment_IN_FLIGHT:
+				continue
+
+			case lnrpc.Payment_SUCCEEDED:
+				return nil, errors.New("warning, the fee " +
+					"estimation payment probe " +
+					"unexpectedly succeeded. Please reach" +
+					"out to the probe destination to " +
+					"negotiate a refund. Otherwise the " +
+					"payment probe amount is lost forever")
+
+			case lnrpc.Payment_FAILED:
+				// Incorrect payment details points to a
+				// successful probe.
+				if payment.FailureReason == lnrpc.PaymentFailureReason_FAILURE_REASON_INCORRECT_PAYMENT_DETAILS { //nolint:lll
```

Comment:
should this check which node in the route returned the error?

Reply:
Will add a check to verify that the index of the hop that returned `incorrect payment details` is the last index of the route. In any other case we'll flag this as an probing error to the client.